### PR TITLE
Add getData event to TagModule

### PR DESCRIPTION
**Pull request by cchabilall83**
In order to allow addons to customize which tags are being displayed, we have to allow them to hook in and modify the data.

### CODE REVIEW

File: applications/vanilla/modules/class.tagmodule.php [hunk 1]
```diff
@@ -103,6 +103,13 @@ public function getData() {
 
         $this->autoContext();
 
+        $this->EventArguments['id'] = $this->ParentID;
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: For the sake of simplicity, I'd keep these names the same. You'll know the ID maps to the ParentID of the tags and not an ID of a tag.

File: applications/vanilla/modules/class.tagmodule.php [hunk 2]
```diff
@@ -103,6 +103,13 @@ public function getData() {
...
+        $this->EventArguments['_TagData'] = &$this->_TagData;
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: I would avoid attempting to give addons a backdoor to directly alter a protected property on the class. You can set a local tag data variable to a null value and pass that to the event. Addons hooking in can check to see if the value is still null before acting, so they don't overwrite each other's data. After all the handlers have been run through, `TagModule::getData` can check to see if addons populated the local variable and set its `_TagData` property as necessary.

File: applications/vanilla/modules/class.tagmodule.php [hunk 3]
```diff
@@ -103,6 +103,13 @@ public function getData() {
...
+        $this->fireEvent('getData');
+        if($this->_TagData) {
+            return;
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: This method has some pretty specific criteria for its data. It should be an instance of `Gdn_DataSet` with a type of `DATASET_TYPE_ARRAY`. Data populated as the result of the getData event should be validated and possibly converted, as necessary, so we're sure what the module is expecting for its data is actually what it's getting.

**Comment by initvector** [comment]
I updated the title and description to be a little more generic and not so specific to Ideation.

File: applications/vanilla/modules/class.tagmodule.php [hunk 4]
```diff
@@ -103,6 +103,15 @@ public function getData() {
 
         $this->autoContext();
 
+        $tagData = null;
+        $this->EventArguments['ParentID'] = $this->ParentID;
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: ParentType is available and used by this method, so it's probably worth passing along too.

**Comment by cchabilall83** [comment]
@initvector this is also related to [PR](https://github.com/vanilla/internal/pull/1552).

File: applications/vanilla/modules/class.tagmodule.php [hunk 5]
```diff
@@ -103,6 +103,15 @@ public function getData() {
...
+        $this->EventArguments['tagData'] = &$tagData;
+        $this->fireEvent('getData');
+        if(is_array($tagData) && $tagData) {
```
Comment (linc) [inline applications/vanilla/modules/class.tagmodule.php]: Coding standard nitpick: add a space after `if`.

File: applications/vanilla/modules/class.tagmodule.php [hunk 6]
```diff
@@ -103,6 +103,15 @@ public function getData() {
 
         $this->autoContext();
```
Comment (linc) [inline applications/vanilla/modules/class.tagmodule.php]: Somewhere in here I feel like an inline comment would be appropriate about why this change is being made. "// Allow an addon to hijack the setting of our data, which aborts the normal database queries." or something to that effect.

File: applications/vanilla/modules/class.tagmodule.php (hunk 5 above)
Comment (initvector) [reply applications/vanilla/modules/class.tagmodule.php]: What if no tags are supposed to be displayed (e.g. the only tag is a "status" tag), so `$tagData` is an empty array (falsy)? In its current state, wouldn't the code continue on to the standard routine?

File: applications/vanilla/modules/class.tagmodule.php [hunk 7]
```diff
@@ -102,6 +102,16 @@ public function getData() {
         $tagQuery = Gdn::sql();
 
         $this->autoContext();
+        //Allow addon to manipulate the data being rendered
+        $tagData = null;
+        $this->EventArguments['ParentID'] = $this->ParentID;
+        $this->EventArguments['tagData'] = &$tagData;
+        $this->fireEvent('getData');
+        if ($tagData !== null) {
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: The expectation is that, after this method is run, *something* will be in the `_TagData` property. The logic here doesn't look like it will allow that. First, we're checking to see if it's `null`. That's good. We're attempting to confirm an addon modified the value.  Next, we're making sure the value is an array. If it is, we're updating the `_TagData` property with the value and exiting the method. If it isn't an array, we're just exiting the method. `_TagData` is never updated with anything.

File: applications/vanilla/modules/class.tagmodule.php [hunk 8]
```diff
@@ -102,6 +102,16 @@ public function getData() {
...
+            if (is_array($tagData))
+            $this->_TagData = new Gdn_DataSet($tagData, DATASET_TYPE_ARRAY);
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: Something about this formatting seems off...

File: applications/vanilla/modules/class.tagmodule.php [hunk 9]
```diff
@@ -102,6 +102,16 @@ public function getData() {
         $tagQuery = Gdn::sql();
 
         $this->autoContext();
+        //Allow addon to manipulate the data being rendered
+        $tagData = null;
+        $this->EventArguments['ParentID'] = $this->ParentID;
+        $this->EventArguments['tagData'] = &$tagData;
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: Please see https://github.com/vanilla/vanilla/pull/7120#discussion_r185049470

File: applications/vanilla/modules/class.tagmodule.php (hunk 7 above)
Comment (initvector) [reply applications/vanilla/modules/class.tagmodule.php]: That definitely wasn't the case [when I commented on it](https://github.com/vanilla/vanilla/blob/22718c1bba70a1df954bea0959a42890ed3a039f/applications/vanilla/modules/class.tagmodule.php#L110).

File: applications/vanilla/modules/class.tagmodule.php [hunk 10]
```diff
@@ -102,6 +102,18 @@ public function getData() {
         $tagQuery = Gdn::sql();
 
         $this->autoContext();
+        //Allow addon to manipulate the data being rendered
+        $tagData = null;
+        $this->EventArguments['ParentID'] = $this->ParentID;
+        $this->EventArguments['ParentID'] = $this->ParentType;
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: Why are you setting ParentID twice?

File: applications/vanilla/modules/class.tagmodule.php [hunk 11]
```diff
@@ -102,6 +102,18 @@ public function getData() {
...
+        $this->EventArguments['tagData'] = &$tagData;
+        $this->fireEvent('getData');
+        if ($tagData !== null) {
+            if (is_array($tagData)) {
```
Comment (initvector) [inline applications/vanilla/modules/class.tagmodule.php]: At this point, since we're checking if it's not null, then checking if it's an array, couldn't we just check to see if it's an array (i.e. drop the is-null condition and bring out the `is_array` check)? If there's a reason to keep the null check, couldn't you combine the condition from the inner block?

Comment (cchabilall83) [reply applications/vanilla/modules/class.tagmodule.php]: I think we drop the is_null check, but I don't think it's necessary
